# **Understand data**

## **1. Chuẩn bị dữ liệu**

In [6]:
# raw data without 2 predict columns
! wget https://github.com/thanggnguyenn/data-for-everything/raw/main/credit_card_churn_data.csv

# raw data on Kaggle
# ! wget https://github.com/thanggnguyenn/data-for-everything/raw/main/credit_card_churn.csv

--2023-03-29 04:25:37--  https://github.com/thanggnguyenn/data-for-everything/raw/main/credit_card_churn.csv
Resolving github.com (github.com)... 140.82.113.3
Connecting to github.com (github.com)|140.82.113.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/thanggnguyenn/data-for-everything/main/credit_card_churn.csv [following]
--2023-03-29 04:25:37--  https://raw.githubusercontent.com/thanggnguyenn/data-for-everything/main/credit_card_churn.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.108.133, 185.199.111.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1510880 (1.4M) [text/plain]
Saving to: ‘credit_card_churn.csv’

credit_card_churn.c 100%[===================>]   1.44M  --.-KB/s    in 0.06s   

2023-03-29 04:25:38 (26.0 MB/s) - ‘credit_card_churn.csv’ s

## **2. Thư viện**

In [1]:
import pandas as pd

## **3. Tải dữ liệu**

In [7]:
raw_data = pd.read_csv("credit_card_churn.csv",index_col="CLIENTNUM")
raw_data.head()

Attrition_Flag  Customer_Age Gender  Dependent_count  \
CLIENTNUM                                                            
768805383  Existing Customer            45      M                3   
818770008  Existing Customer            49      F                5   
713982108  Existing Customer            51      M                3   
769911858  Existing Customer            40      F                4   
709106358  Existing Customer            40      M                3   

          Education_Level Marital_Status Income_Category Card_Category  \
CLIENTNUM                                                                
768805383     High School        Married     $60K - $80K          Blue   
818770008        Graduate         Single  Less than $40K          Blue   
713982108        Graduate        Married    $80K - $120K          Blue   
769911858     High School        Unknown  Less than $40K          Blue   
709106358      Uneducated        Married     $60K - $80K          Blue   

           Months_on_book  Total_Relationship_Count  ...  Credit_Limit  \
CLIENTNUM                                            ...                 
768805383              39                         5  ...       12691.0   
818770008              44                         6  ...        8256.0   
713982108              36                         4  ...        3418.0   
769911858              34                         3  ...        3313.0   
709106358              21                         5  ...        4716.0   

           Total_Revolving_Bal  Avg_Open_To_Buy  Total_Amt_Chng_Q4_Q1  \
CLIENTNUM                                                               
768805383                  777          11914.0                 1.335   
818770008                  864           7392.0                 1.541   
713982108                    0           3418.0                 2.594   
769911858                 2517            796.0                 1.405   
709106358                    0           4716.0                 2.175   

           Total_Trans_Amt  Total_Trans_Ct  Total_Ct_Chng_Q4_Q1  \
CLIENTNUM                                                         
768805383             1144              42                1.625   
818770008             1291              33                3.714   
713982108             1887              20                2.333   
769911858             1171              20                2.333   
709106358              816              28                2.500   

           Avg_Utilization_Ratio  \
CLIENTNUM                          
768805383                  0.061   
818770008                  0.105   
713982108                  0.000   
769911858                  0.760   
709106358                  0.000   

           Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1  \
CLIENTNUM                                                                                                                                       
768805383                                           0.000093                                                                                    
818770008                                           0.000057                                                                                    
713982108                                           0.000021                                                                                    
769911858                                           0.000134                                                                                    
709106358                                           0.000022                                                                                    

           Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2  
CLIENTNUM                                                                                 

In [9]:
raw_data.shape

(10127, 22)

Dữ liệu có 10127 dòng và 22 cột.

## **4. Tổng quan về dữ liệu** 

### **Mô tả bộ dữ liệu**

1. CLIENTNUM: số định danh khách hàng có tài khoản tại ngân hàng. 

2. Attrition_Flag: Hoạt động của khách hàng. Gồm hai giá trị là: Existing Customer (khách hàng đang hoạt động) và Attrited Customer (khách hàng đã ngưng sử dụng dịch vụ). 

3. Customer_Age: tuổi của khách hàng. 

4. Gender: giới tính 

5. Dependent_count: số người phụ thuộc. 

6. Education_Level: trình độ học vấn. 

7. Marital_Status: tình trạng hôn nhân. 

8. Income_Category: phân loại mức thu nhập (gồm có: dưới \$40K, \$40K – \$60K, \$60K - \$80K, \$80K - \$120K, trên \$129K và không biết).  

9. Card_Category: loại thẻ tín dụng (Blue, Silver, Gold, Platinum). 

10. Months_on_book: thời gian có giao dịch với ngân hàng. 

11. Total_Relationship_Count: số lượng toàn bộ sản phẩm mà khách hàng đã sử dụng. 

12. Months_Inactive_12_mon: số tháng không giao dịch trong 12 tháng gần nhất. 

13. Contacts_Count_12_mon: số liên lạc trong 12 tháng gần nhất. 

14. Credit_Limit: hạn mức tín dụng trên thẻ tín dụng. 

15. Total_Revolving_Bal: tổng số dư quay vòng trên thẻ tín dụng.

16. Avg_Open_To_Buy: mức hạn tín dụng còn lại (trung bình của 12 tháng gần nhất).

17. Total_Amt_Chng_Q4_Q1: thay đổi trong số tiền giao dịch. 

18. Total_Trans_Amt: tổng số tiền giao dịch (trong 12 tháng gần nhất). 

19. Total_Trans_Ct: tổng số lượng giao dịch (trong 12 tháng gần nhất). 

20. Total_Ct_Chng_Q4_Q1: thay đổi trong số lượng giao dịch. 

21. Avg_Utilization_Ratio: tỷ lệ sử dụng thẻ trung bình. 

22. Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1: mô hình Naïve Bayes dự đoán khách hàng có ngưng sử dụng dịch vụ dựa trên các yếu tố sau. (không sử dụng cột này). 

23. Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2: mô hình Naïve Bayes dự đoán khách hàng có ngưng sử dụng dịch vụ dựa trên các yếu tố sau. (không sử dụng cột này). 

**Nhận xét:** chúng ta sẽ xóa hai cột 22 và 23 do không sử dụng hai cột này để xây dựng mô hình. 

In [8]:
data1 = raw_data.drop(columns=\
["Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1",\
"Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2"],\
    )
data1.head()

Attrition_Flag  Customer_Age Gender  Dependent_count  \
CLIENTNUM                                                            
768805383  Existing Customer            45      M                3   
818770008  Existing Customer            49      F                5   
713982108  Existing Customer            51      M                3   
769911858  Existing Customer            40      F                4   
709106358  Existing Customer            40      M                3   

          Education_Level Marital_Status Income_Category Card_Category  \
CLIENTNUM                                                                
768805383     High School        Married     $60K - $80K          Blue   
818770008        Graduate         Single  Less than $40K          Blue   
713982108        Graduate        Married    $80K - $120K          Blue   
769911858     High School        Unknown  Less than $40K          Blue   
709106358      Uneducated        Married     $60K - $80K          Blue   

           Months_on_book  Total_Relationship_Count  Months_Inactive_12_mon  \
CLIENTNUM                                                                     
768805383              39                         5                       1   
818770008              44                         6                       1   
713982108              36                         4                       1   
769911858              34                         3                       4   
709106358              21                         5                       1   

           Contacts_Count_12_mon  Credit_Limit  Total_Revolving_Bal  \
CLIENTNUM                                                             
768805383                      3       12691.0                  777   
818770008                      2        8256.0                  864   
713982108                      0        3418.0                    0   
769911858                      1        3313.0                 2517   
709106358                      0        4716.0                    0   

           Avg_Open_To_Buy  Total_Amt_Chng_Q4_Q1  Total_Trans_Amt  \
CLIENTNUM                                                           
768805383          11914.0                 1.335             1144   
818770008           7392.0                 1.541             1291   
713982108           3418.0                 2.594             1887   
769911858            796.0                 1.405             1171   
709106358           4716.0                 2.175              816   

           Total_Trans_Ct  Total_Ct_Chng_Q4_Q1  Avg_Utilization_Ratio  
CLIENTNUM                                                              
768805383              42                1.625                  0.061  
818770008              33                3.714                  0.105  
713982108              20                2.333                  0.000  
769911858              20                2.333                  0.760  
709106358              28                2.500                  0.000

In [10]:
data1.shape

(10127, 20)

Sau khi xóa hai cột cuối thì dữ liệu còn 20 thuộc tính.

In [11]:
# lưu data có 20 cột
# data1.to_csv("credit_card_churn_20_cols.csv")

## **5. Tìm hiểu về dữ liệu**

In [ ]:
data = pd.read_csv("credit_card_churn_20_cols.csv",index_col="CLIENTNUM")
data.head()